### PR TITLE
chore: pipeline health checks during rollout

### PR DESCRIPTION
**Pull request by jon-funk**
# Description

This new scritpt+action runs in tandem with the helm build to QA it at the infra level, highlighting issues that helm may not report back about. Additionally, this script is configured to timeout just before helm and collect information in the event of a failing/timing out helm build, so you get a report on what might be wrong in-PR.

Example of it catching a probe failure in events: https://github.com/bcgov/nr-compliance-enforcement/actions/runs/12170089403/job/33944486709?pr=795

Example of an all green success: https://github.com/bcgov/nr-compliance-enforcement/actions/runs/12171219493/job/33947826812

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [X] Demonstrated success and failure examples

## Checklist

- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Further comments

---

Thanks for the PR!

Deployments, as required, will be available below:
- [Frontend](https://nr-compliance-enforcement-795-frontend.apps.silver.devops.gov.bc.ca)
- [Backend](https://nr-compliance-enforcement-795-frontend.apps.silver.devops.gov.bc.ca/api)

Please create PRs in draft mode.  Mark as ready to enable:
- [Analysis Workflow](https://github.com/bcgov/nr-compliance-enforcement/actions/workflows/analysis.yml)

After merge, new images are deployed in:
- [Merge Workflow](https://github.com/bcgov/nr-compliance-enforcement/actions/workflows/merge.yml)

---

Thanks for the PR!

Deployments, as required, will be available below:
- [Frontend](https://nr-compliance-enforcement-795-frontend.apps.silver.devops.gov.bc.ca)
- [Backend](https://nr-compliance-enforcement-795-frontend.apps.silver.devops.gov.bc.ca/api)

Please create PRs in draft mode.  Mark as ready to enable:
- [Analysis Workflow](https://github.com/bcgov/nr-compliance-enforcement/actions/workflows/analysis.yml)

After merge, new images are deployed in:
- [Merge Workflow](https://github.com/bcgov/nr-compliance-enforcement/actions/workflows/merge.yml)

### CODE REVIEW

File: charts/app/templates/backend/templates/deployment.yaml [hunk 1]
```diff
@@ -22,6 +22,7 @@ spec:
       labels:
         {{- include "backend.labels" . | nindent 8 }}
     spec:
+      minReadySeconds: 10
```
Comment (mishraomp) [inline charts/app/templates/backend/templates/deployment.yaml:25]: do we need to add extra 10 seconds delay for pod to be ready to serve incoming traffic .

Comment (jon-funk) [reply charts/app/templates/backend/templates/deployment.yaml:25]: We've had some api errors from fresh pods starting up in test when under load. I have another PR that lightens up the probes and scaling rates from the HPA - but the init container timing is tricky with probes in general. My thinking here was just a bit of extra security, although now that you mention it 10s may be too much. Thoughts?

Comment (mishraomp) [reply charts/app/templates/backend/templates/deployment.yaml:25]: I would recommend we fix the readiness probe, if required, we add startup probes, but once pod is ready, waiting for 10 seconds, sounds not ideal to me

Comment (jon-funk) [reply charts/app/templates/backend/templates/deployment.yaml:25]: I should mention that on the application level, our health check on the api backend is 'dumb' in that it just responds to the http request, not actually if the db is connected and happy, so the k8s probe's source of truth is unreliable, so I'm compensating for that on the infra level. For now.

Comment (jon-funk) [reply charts/app/templates/backend/templates/deployment.yaml]: removed these - appreciate the feedback @mishraomp

Comment (mishraomp) [reply charts/app/templates/backend/templates/deployment.yaml]: > removed these - appreciate the feedback @mishraomp

Thanks Jon, I would vote for another ticket in the backlog to fix the readiness and liveness probe in the API :)

File: .github/scripts/rollout_healthcheck.sh [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+    for deployment in $deployment_list; do
+        deployment_status=$(_health_check_deployment $deployment)
+        if [ "$deployment_status" == "false" ]; then
+            succeeded="false"
+            break
+        fi
+    done
+    for pod in $pod_list; do
+        _pod_running=$(_pod_running $pod)
+        if [ "$pod_status" == "false" ]; then
```
Comment (afwilcox) [inline .github/scripts/rollout_healthcheck.sh:173]: Is this correct?  Is pod_status available in this context when it's deployed as a local variable in the _pod_running method?

Why not just if [ " (_pod_running $pod)" == "false ]

File: .github/scripts/rollout_healthcheck.sh [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+
+# global flag to indicate if health check passed
+HEALTH_CHECK_PASSED=0
+
+
+set -e # failfast
+trap 'echo "Error occurred at line $LINENO while executing function $FUNCNAME"' ERR
+
+help_str() {
+    echo "Usage: SKIP_AUTH=true LABEL_SELECTOR=\"app.kubernetes.io/instance=nr-compliance-enforcement-PRNUM\" OC_NAMESPACE=c1c7ed-dev .github/scripts/rollout_healthcheck.sh"
```
Comment (afwilcox) [inline .github/scripts/rollout_healthcheck.sh:50]: Should the script also contain checks for LABEL_SELECTOR and OC_NAMESPACE to validate they are passed in?

File: .github/scripts/rollout_healthcheck.sh [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+
+# will be set to 1 if a timeout occurs from an unfinished rollout
+# we use this to fetch triage info
+TIMED_OUT=0
+
+# string of commands to run after health check for additional info
+COMMANDS_TO_RUN=""
+
+# global flag to indicate if health check passed
+HEALTH_CHECK_PASSED=0
```
Comment (afwilcox) [inline .github/scripts/rollout_healthcheck.sh:43]: This seems counter-intuitive.    It looks like this script sets the value "HEALTH_CHECK_PASSED" = 1 (true) whenever a failure condition is encountered.  

Maybe rename HEALTH_CHECK_FAILED?

**Review by afwilcox — CHANGES_REQUESTED** [review]

**Review by afwilcox — APPROVED** [review]

**Pull request merged**